### PR TITLE
implement parameter qsort

### LINKED ISSUE #13: Parameter -qsort

### PR DESCRIPTION
**Pull request by justinmueller05082000**
fixes #13

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ func main() {
 	ignoreCaseParameter := flag.Bool("f", false, "Fold lower case to upper case characters")
 	randomSortingParameter := flag.Bool("R", false, "shuffle, but group identical keys.")
 	reversePrintingParameter := flag.Bool("r", false, "Print result in reverse order")
+	quickSortParameter := flag.Bool("qsort", false, "Use quick sort, if the sort specifications allow")
```
Comment (Al2Klimov) [inline main.go]: Which "sort specifications"?

File: main.go [hunk 2]
```diff
@@ -88,17 +89,23 @@ func main() {
...
 
-	for i := 0; i < len(dataSplit); i++ {
-		j := i
-		for j > 0 && comparator(dataSplit[j], dataSplit[j-1]) < 0 {
-			dataSplit[j], dataSplit[j-1] = dataSplit[j-1], dataSplit[j]
-			j -= 1
+	if *quickSortParameter {
+		quickSort(dataSplit)
+	} else {
+		for i := 0; i < len(dataSplit); i++ {
```
Comment (Al2Klimov) [inline main.go]: Either all or none shall be outsourced into functions.

File: main.go [hunk 3]
```diff
@@ -88,17 +89,23 @@ func main() {
...
+			j := i
+			for j > 0 && comparator(dataSplit[j], dataSplit[j-1]) < 0 {
+				dataSplit[j], dataSplit[j-1] = dataSplit[j-1], dataSplit[j]
+				j -= 1
+			}
 		}
+
```
Comment (Al2Klimov) [inline main.go]: One line too many.

File: main.go [hunk 4]
```diff
@@ -131,3 +138,27 @@ func wrapReverse(f func(a, b []byte) int) func(a, b []byte) int {
...
+	if len(input) < 2 {
+		return input
+	}
+
+	left, right := 0, len(input)-1
+	pivot := rand.Int() % len(input)
+	input[pivot], input[right] = input[right], input[pivot]
+
+	for i := range input {
+		if bytes.Compare(input[i], input[right]) < 0 {
```
Comment (Al2Klimov) [inline main.go]: What about -n and -R?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (justinmueller05082000) [reply main.go]: i dont know i just copied the original explanation

File: main.go (hunk 3 above)
Comment (justinmueller05082000) [reply main.go]: removed

File: main.go (hunk 1 above)
Comment (Al2Klimov) [reply main.go]: Yes, you're more or less re-writing the original program, but the help is about *your new program*. I.e. if there aren't any "sort specifications" in *your new program*, you shall not write about them in the help text.

File: main.go (hunk 2 above)
Comment (justinmueller05082000) [reply main.go]: Am I wrong with putting my sort algo into a function?

Comment (Al2Klimov) [reply main.go]: No, you're wrong with putting some in functions and some not.

Comment (justinmueller05082000) [reply main.go]: I did it like that but i guess I misunderstood the exercise.
```
if *quickSortParameter {
		quickSort(dataSplit)
	} else {
		bubbleSort(dataSplit, comparator)
	}

```

```
func bubbleSort(input [][]byte, comp func(a, b []byte) int) [][]byte {
	for i := 0; i < len(input); i++ {
		j := i
		for j > 0 && comp(input[j], input[j-1]) < 0 {
			input[j], input[j-1] = input[j-1], input[j]
			j -= 1
		}
	}
	return input
}
```

File: main.go [hunk 5]
```diff
@@ -88,17 +89,16 @@ func main() {
...
+		quickSort(dataSplit, comparator)
+	} else {
+		bubbleSort(dataSplit, comparator)
 	}
+
 	dataJoin := bytes.Join(dataSplit, []byte{'\n'})
 
 	if *writingParameter != "" {
-		writeErr := ioutil.WriteFile(*writingParameter, append(dataJoin,'\n'), 0644)
+		writeErr := ioutil.WriteFile(*writingParameter, append(dataJoin, '\n'), 0644)
```
Comment (Al2Klimov) [inline main.go]: Not part of the task.

File: main.go [hunk 6]
```diff
@@ -131,3 +131,38 @@ func wrapReverse(f func(a, b []byte) int) func(a, b []byte) int {
...
+
+func bubbleSort(input [][]byte, comp func(a, b []byte) int) [][]byte {
+	for i := 0; i < len(input); i++ {
+		j := i
+		for j > 0 && comp(input[j], input[j-1]) < 0 {
+			input[j], input[j-1] = input[j-1], input[j]
+			j -= 1
+		}
+	}
+	return input
```
Comment (Al2Klimov) [inline main.go:168]: I'd separate the loop and the return by an empty line.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
* [x] `git reset --soft HEAD~$(git rev-list --count $(git merge-base HEAD master)..HEAD^)`
* [x] `git commit --amend --no-edit`
* [x] `git push -f origin feature/parameter-qsort-13`

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol w/ and w/o -qsort and w/ and w/o -n (2 x 2). You don't need to compare with the original sort if the Linux version doesn't have -qsort.

**Comment by justinmueller05082000** [comment]
**Test protocol:**

**toSort.txt:**
```
10
12
425
6
3
5
66
3
506
39
43
```

**Case 1:**
**Terminal:** `go run . toSort.txt `
**Output:**
```
10
12
3
3
39
425
43
5
506
6
66
```


**Case 2:**
**Terminal:** `go run . -qsort toSort.txt `
**Output:**
```
10
12
3
3
39
425
43
5
506
6
66
```


**Case 3;**
**Terminal:** `go run . -qsort -n toSort.txt `
**Output:**
```
3
3
5
6
10
12
39
43
66
425
506
```


**Case 4:**
**Terminal:**` go run . -n toSort.txt `
**Output:**
```
3
3
5
6
10
12
39
43
66
425
506
```

**Conclusion:**
Qsort prints the same result as our normal sort.

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**